### PR TITLE
fix(core): accessibility fix for multi-input component

### PR DESCRIPTION
**Pull request by DeepakSap14**
#### Please provide a link to the associated issue.
Closes https://github.com/SAP/fundamental-ngx/issues/5578

#### Please provide a brief summary of this pull request.
fixes accessibility issues in core multi-input. 

#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/main/CONTRIBUTING.md
- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist
- [x] Run npm run build-pack-library and test in external application

Documentation checklist:
- [NA] update `README.md`
- [NA] [Breaking Changes Wiki](https://github.com/SAP/fundamental-ngx/wiki/Breaking-Changes)
- [x] Documentation Examples
- [x] Stackblitz works for all examples

### CODE REVIEW

**Comment by DeepakSap14** [comment]
fix pending for issue #4, #6, #8.
other issues are fixed.

**Comment by DeepakSap14** [comment]
will reopen once other PR's are merged. ETA 1-2 days.

File: apps/docs/src/app/core/component-docs/multi-input/examples/multi-input-async-example/multi-input-async-example.component.html [hunk 1]
```diff
@@ -1,4 +1,7 @@
+<label fd-label for="multiInputAsync">Multi Input with async data</label>
 <fd-multi-input
+    inputId="multiInputAsync"
+    [title]="'search'"
```
Comment (DeepakSap14) [inline apps/docs/src/app/core/component-docs/multi-input/examples/multi-input-async-example/multi-input-async-example.component.html:4]: doing `title='search'` will show `tooltip` on input as well, which is not needed. this title input is to show `tooltip` on add-on button only. 
It is causing issue while screen reading. so have to bind here.

**Review by Betrozov — APPROVED** [review]

File: libs/core/src/lib/multi-input/multi-input-mobile/multi-input-mobile.component.html [hunk 2]
```diff
@@ -5,6 +5,9 @@ <h1 fd-title *ngIf="mobileConfig?.title">{{ mobileConfig?.title }}</h1>
             <button fd-dialog-close-button
                     [mobile]="true"
                     *ngIf="mobileConfig?.hasCloseButton"
+                    title="close"
+                    i18n-title="@@coreI18nMobileMultiInputClose"
+                    tabindex="0"
```
[the remark this answers is not in the document]
Comment (InnaAtanasova) [reply libs/core/src/lib/multi-input/multi-input-mobile/multi-input-mobile.component.html:10]: Good question. We are starting to abuse the tabindex. Elements that naturally gets focus should not have tabindex set.

File: libs/core/src/lib/multi-input/multi-input.component.ts [hunk 3]
```diff
@@ -19,27 +20,32 @@ import {
...
+import { MenuKeyboardService } from '@fundamental-ngx/core/menu';
+import { MobileModeConfig } from '@fundamental-ngx/core/mobile-mode';
+import { PopoverComponent } from '@fundamental-ngx/core/popover';
+import { FormStates, PopoverFillMode } from '@fundamental-ngx/core/shared';
+import { TokenizerComponent } from '@fundamental-ngx/core/token';
+
 import { MultiInputMobileComponent } from './multi-input-mobile/multi-input-mobile.component';
 import { MULTI_INPUT_COMPONENT, MultiInputInterface } from './multi-input.interface';
 
+let inputRandomId = 0;
```
Comment (InnaAtanasova) [inline libs/core/src/lib/multi-input/multi-input.component.ts:47]: It's actually not `random` but `unique`. Please be consistent with the other components in the library: 
Screen Shot 2021-08-09 at 2 37 33 PM

File: libs/core/src/lib/multi-input/multi-input.component.ts [hunk 4]
```diff
@@ -111,7 +110,7 @@ export class MultiInputComponent implements
 
     /** Id attribute for input element inside MultiInput component */
     @Input()
-    inputId = '';
+    inputId = `fd-input-id-${inputRandomId++}`;
```
Comment (InnaAtanasova) [inline libs/core/src/lib/multi-input/multi-input.component.ts:113]: inputId = `fd-input-${inputRandomId++}`;

File: libs/core/src/lib/multi-input/multi-input.component.ts [hunk 5]
```diff
@@ -219,6 +218,23 @@ export class MultiInputComponent implements
...
+     * Second part of message for search suggestion.
+     * direction for navigating the suggestion. This is not necessry in case of 0 suggestion.
+     */
+    @Input()
+    searchSuggestionNavigateMessage = 'use up and down arrows to navigate';
+
+    /** Message for allowing token from values available */
+    @Input()
+    tokenSelectMessage = 'select value from the list';
+
```
Comment (InnaAtanasova) [inline libs/core/src/lib/multi-input/multi-input.component.ts:237]: These being set by default, how would they be translated in different languages?

File: libs/core/src/lib/multi-input/multi-input.component.ts [hunk 6]
```diff
@@ -335,6 +357,17 @@ export class MultiInputComponent implements
         this._subscriptions.unsubscribe();
     }
 
+    /** @hidden */
+    @HostListener('keydown', ['$event'])
+    keyDown(keyboardEvent: KeyboardEvent): void {
```
Comment (InnaAtanasova) [inline libs/core/src/lib/multi-input/multi-input.component.ts:362]: missing prefix for hidden method

**Review by InnaAtanasova — DISMISSED** [review]

File: libs/core/src/lib/multi-input/multi-input-mobile/multi-input-mobile.component.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (DeepakSap14) [reply libs/core/src/lib/multi-input/multi-input-mobile/multi-input-mobile.component.html]: removed tabindex. without it also, focus is going on to close button.

File: apps/docs/src/app/core/component-docs/multi-input/examples/multi-input-mobile-example/multi-input-mobile-example.component.html [hunk 7]
```diff
@@ -1,5 +1,7 @@
-<label fd-form-label>Mobile Mode Multi Input</label>
+<label fd-form-label for="mobilemultiInput">Mobile Mode Multi Input</label>
 <fd-multi-input [mobile]="true"
+                inputId="mobilemultiInput"
```
[the remark this answers is not in the document]
Comment (DeepakSap14) [reply apps/docs/src/app/core/component-docs/multi-input/examples/multi-input-mobile-example/multi-input-mobile-example.component.html:3]: Done

File: apps/docs/src/app/core/component-docs/multi-input/examples/multi-input-form-example/multi-input-form-example.component.html [hunk 8]
```diff
@@ -12,8 +14,10 @@
             </fd-multi-input>
         </div>
         <div fd-form-item>
-            <label fd-form-label>Disabled Multi Input</label>
+            <label fd-form-label for="disabledmultiInput">Disabled Multi Input</label>
```
[the remark this answers is not in the document]
Comment (DeepakSap14) [reply apps/docs/src/app/core/component-docs/multi-input/examples/multi-input-form-example/multi-input-form-example.component.html]: Done

File: apps/docs/src/app/core/component-docs/multi-input/examples/multi-input-form-example/multi-input-form-example.component.html [hunk 9]
```diff
@@ -1,8 +1,10 @@
 <form [formGroup]="customForm">
     <fieldset fd-fieldset>
         <div fd-form-item>
-            <label fd-form-label required="true">Valid Multi Input</label>
+            <label fd-form-label for="validmultiInput" required="true">Valid Multi Input</label>
```
[the remark this answers is not in the document]
Comment (DeepakSap14) [reply apps/docs/src/app/core/component-docs/multi-input/examples/multi-input-form-example/multi-input-form-example.component.html]: Done

File: libs/core/src/lib/multi-input/multi-input.component.ts [hunk 10]
```diff
@@ -263,9 +279,14 @@ export class MultiInputComponent implements
     @ViewChild(TokenizerComponent)
     tokenizer: TokenizerComponent;
 
+    @ViewChild('inputfield', { read: ElementRef })
+    inputfield: ElementRef;
+
     /** @hidden */
     displayedValues: any[] = [];
 
+    currentSearchSuggestionAnnoucementMessage = '';
```
[the remark this answers is not in the document]
Comment (DeepakSap14) [reply libs/core/src/lib/multi-input/multi-input.component.ts]: Done

File: libs/core/src/lib/multi-input/multi-input.component.ts [hunk 11]
```diff
@@ -263,9 +279,14 @@ export class MultiInputComponent implements
     @ViewChild(TokenizerComponent)
     tokenizer: TokenizerComponent;
 
+    @ViewChild('inputfield', { read: ElementRef })
+    inputfield: ElementRef;
```
[the remark this answers is not in the document]
Comment (DeepakSap14) [reply libs/core/src/lib/multi-input/multi-input.component.ts]: Done

File: libs/core/src/lib/multi-input/multi-input.component.ts [hunk 12]
```diff
@@ -423,6 +456,9 @@ export class MultiInputComponent implements
         this.tokenizer.removeSelectedTokens();
 
         this._changeDetRef.detectChanges();
+
+        // focus on input when mobile dialog opens
+        this.inputfield.nativeElement.focus();
```
[the remark this answers is not in the document]
Comment (DeepakSap14) [reply libs/core/src/lib/multi-input/multi-input.component.ts]: verified in Chrome, firefox, safari, edge.
it is focusing on `input field` when mobile dialog opens.

mobile-mode-multiInput

File: libs/core/src/lib/multi-input/multi-input.component.ts (hunk 3 above)
Comment (DeepakSap14) [reply libs/core/src/lib/multi-input/multi-input.component.ts]: Done

File: libs/core/src/lib/multi-input/multi-input.component.ts (hunk 4 above)
Comment (DeepakSap14) [reply libs/core/src/lib/multi-input/multi-input.component.ts]: Done

File: libs/core/src/lib/multi-input/multi-input.component.ts (hunk 5 above)
Comment (DeepakSap14) [reply libs/core/src/lib/multi-input/multi-input.component.ts]: using `localize` now. it is generating message for translation now.

Screenshot 2021-08-10 at 12 34 11 PM

File: libs/core/src/lib/multi-input/multi-input.component.ts (hunk 6 above)
Comment (DeepakSap14) [reply libs/core/src/lib/multi-input/multi-input.component.ts]: Done

File: libs/core/src/lib/multi-input/multi-input.component.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (kavya-b) [reply libs/core/src/lib/multi-input/multi-input.component.ts]: i'm sorry i should have clarified the steps. It happens when i click on the value help icon at the right and open the mobile dialog

File: apps/docs/src/app/core/component-docs/multi-input/examples/multi-input-form-example/multi-input-form-example.component.html [hunk 13]
```diff
@@ -1,8 +1,10 @@
 <form [formGroup]="customForm">
     <fieldset fd-fieldset>
         <div fd-form-item>
-            <label fd-form-label required="true">Valid Multi Input</label>
+            <label fd-form-label for="validMultiInput" required="true">Valid Multi Input</label>
             <fd-multi-input
+                inputId="validmultiInput"
```
Comment (kavya-b) [inline apps/docs/src/app/core/component-docs/multi-input/examples/multi-input-form-example/multi-input-form-example.component.html:6]: camel casing missed for the associated ids in this file, the `for` and `id` should have the same id

File: apps/docs/src/app/core/component-docs/multi-input/examples/multi-input-mobile-example/multi-input-mobile-example.component.html (hunk 7 above)
[the remark this answers is not in the document]
Comment (kavya-b) [reply apps/docs/src/app/core/component-docs/multi-input/examples/multi-input-mobile-example/multi-input-mobile-example.component.html:3]: it is done for `for` but not for the `inputId`

File: libs/core/src/lib/multi-input/multi-input.component.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (DeepakSap14) [reply libs/core/src/lib/multi-input/multi-input.component.ts]: this is fine then, i think. we are opening the value help. so value help list should be in focus. thus first item list gets focus.
I think it is correct behaviour.

in this line, fix is for `when input was entered and mobile mode was opening, focus was not staying on input.`

[the remark this answers is not in the document]
Comment (kavya-b) [reply libs/core/src/lib/multi-input/multi-input.component.ts]: okay, thanks for clarifying!

File: apps/docs/src/app/core/component-docs/multi-input/examples/multi-input-form-example/multi-input-form-example.component.html (hunk 13 above)
Comment (DeepakSap14) [reply apps/docs/src/app/core/component-docs/multi-input/examples/multi-input-form-example/multi-input-form-example.component.html]: sorry, missed it. fixed now.

File: apps/docs/src/app/core/component-docs/multi-input/examples/multi-input-mobile-example/multi-input-mobile-example.component.html (hunk 7 above)
[the remark this answers is not in the document]
Comment (DeepakSap14) [reply apps/docs/src/app/core/component-docs/multi-input/examples/multi-input-mobile-example/multi-input-mobile-example.component.html]: sorry, missed it. fixed now.

**Review by kavya-b — APPROVED** [review]

**Review by mikerodonnell89 — COMMENTED** [review]
@DeepakSap14 see my comment here regarding `$localize` in the library: https://github.com/SAP/fundamental-ngx/pull/6196#discussion_r686189050

**Comment by InnaAtanasova** [comment]
Please check this thread https://github.com/SAP/fundamental-ngx/pull/6196#discussion_r688719078

**Comment by platon-rov** [comment]
Here is the info about `$localize`.

https://github.com/SAP/fundamental-ngx/pull/6196#discussion_r692196402

**Comment by platon-rov** [comment]
Please rebase onto `main` and resolve conflicts.

File: apps/docs/src/app/core/component-docs/multi-input/examples/multi-input-async-example/multi-input-async-example.component.html (hunk 1 above)
Comment (platon-rov) [reply apps/docs/src/app/core/component-docs/multi-input/examples/multi-input-async-example/multi-input-async-example.component.html:4]: I guess in that case we need to write about that fact in the typeDoc for the property.

File: libs/core/src/lib/multi-input/multi-input.component.html [hunk 14]
```diff
@@ -100,7 +103,7 @@
         <li
             *ngFor="let value of displayedValues"
             fd-list-item
-            (click)="openChangeHandle(false)"
+            (click)="!mobile ? openChangeHandle(false): openChangeHandle(true)"
```
Comment (platon-rov) [inline libs/core/src/lib/multi-input/multi-input.component.html:106]: Can be shortened into `openChangeHandle(!!mobile)`.

File: libs/core/src/lib/token/token.component.html [hunk 15]
```diff
@@ -12,5 +15,5 @@
     <span class="fd-token__text no-text-select">
         <ng-content></ng-content>
     </span>
-    <span (click)="closeClickHandler($event)" *ngIf="!readOnly" class="fd-token__close" tabindex="-1"></span>
+    <span (click)="closeClickHandler($event)" *ngIf="!readOnly" class="fd-token__close" title="deletable" tabindex="-1"></span>
```
Comment (platon-rov) [inline libs/core/src/lib/token/token.component.html:18]: i18n for the `title`?

**Review by platon-rov — CHANGES_REQUESTED** [review]

File: libs/core/src/lib/multi-input/multi-input.component.ts (hunk 5 above)
Comment (DeepakSap14) [reply libs/core/src/lib/multi-input/multi-input.component.ts]: following the final discussion. All these value should be a user input.

File: libs/core/src/lib/multi-input/multi-input.component.html (hunk 14 above)
Comment (DeepakSap14) [reply libs/core/src/lib/multi-input/multi-input.component.html]: done

File: libs/core/src/lib/token/token.component.html (hunk 15 above)
Comment (DeepakSap14) [reply libs/core/src/lib/token/token.component.html]: done here and missing places.

**Comment by DeepakSap14** [comment]
closing this PR, will reopen once list and dialog component accessibilities are addressed

**Pull request closed without merging**